README
	What is yespower?

yespower is a proof-of-work (PoW) focused fork of yescrypt.  While
yescrypt is a password-based key derivation function (KDF) and password
hashing scheme, and thus is meant for processing passwords, yespower is
meant for processing trial inputs such as block headers (including
nonces) in PoW-based blockchains.

On its own, yespower isn't a complete proof-of-work system.  Rather, in
the blockchain use case, yespower's return value is meant to be checked
for being numerically no greater than the blockchain's current target
(which is related to mining difficulty) or else the proof attempt
(yespower invocation) is to be repeated (with a different nonce) until
the condition is finally met (allowing a new block to be mined).  This
process isn't specific to yespower and isn't part of yespower itself
(rather, it is similar in many PoW-based blockchains and is to be
defined and implemented externally to yespower) and thus isn't described
in here any further.


	Why or why not yespower?

Different proof-of-work schemes in existence vary in many aspects,
including in friendliness to different types of hardware.  There's
demand for all sorts of hardware (un)friendliness in those - for
different use cases and by different communities.

yespower in particular is designed to be CPU-friendly, GPU-unfriendly,
and FPGA/ASIC-neutral.  In other words, it's meant to be relatively
efficient to compute on current CPUs and relatively inefficient on
current GPUs.  Unfortunately, being GPU-unfriendly also means that
eventual FPGA and ASIC implementations will only compete with CPUs, and
at least ASICs will win over the CPUs (FPGAs might not because of this
market's peculiarities - large FPGAs are even more "over-priced" than
large CPUs are), albeit by far not to the extent they did e.g. for
Bitweb and Litecoin.

There's a lot of talk about "ASIC resistance".  What is (or should be)
meant by that is limiting the advantage of specialized ASICs.  While
limiting the advantage at KDF to e.g. 10x and at password hashing to
e.g. 100x (talking orders of magnitude here, in whatever terms) may be
considered "ASIC resistant" (as compared to e.g. 100,000x we'd have
without trying), similar improvement factors are practically not "ASIC
resistant" for cryptocurrency mining where they can make all the
difference between CPU mining being profitable and not.  There might
also exist in-between PoW use cases where moderate ASIC advantage is OK,
such as with non-cryptocurrency and/or private/permissioned blockchains.

Thus, current yespower may be considered either a short-term choice
(valid until one of its uses provides sufficient perceived incentive to
likely result in specialized ASICs) or a deliberate choice of a pro-CPU,
anti-GPU, moderately-pro-ASIC PoW scheme.  It is also possible to
respond to known improvements in future GPUs/implementations and/or to
ASICs with new versions of yespower that users would need to switch to.


	yespower versions.

yespower includes optimized and specialized re-implementation of the
obsolete yescrypt 0.5 (based off its first submission to Password
Hashing Competition back in 2014) now re-released as yespower 0.5, and
brand new proof-of-work specific variation known as yespower 1.0.

yespower 0.5 is intended as a compatible upgrade for cryptocurrencies
that already use yescrypt 0.5 (providing a few percent speedup), and
yespower 1.0 may be used as a further upgrade or a new choice of PoW by
those and other cryptocurrencies and other projects.

There are many significant differences between yespower 0.5 and 1.0
under the hood, but the main user visible difference is yespower 1.0
greatly improving on GPU-unfriendliness in light of improvements seen in
modern GPUs (up to and including NVIDIA Volta) and GPU implementations
of yescrypt 0.5.  This is achieved mostly through greater use of CPUs'
L2 cache.

The version of algorithm to use is requested through parameters,
allowing for both algorithms to co-exist in client and miner
implementations (such as in preparation for a cryptocurrency hard-fork
and/or supporting multiple cryptocurrencies in one program).


	Parameter selection.

For new uses of yespower, set the requested version to the highest
supported, and set N*r to the highest you can reasonably afford in terms
of proof verification time (which might in turn be determined by desired
share rate per mining pool server), using one of the following options:

1 MiB: N = 1024, r = 8
2 MiB: N = 2048, r = 8
4 MiB: N = 1024, r = 32
8 MiB: N = 2048, r = 32
16 MiB: N = 4096, r = 32

and so on for higher N keeping r=32.

You may also set the personalization string to your liking, but that is
not required (you can set its pointer to NULL and its length to 0).  Its
support is provided mostly for compatibility with existing modifications
of yescrypt 0.5.


	Performance.

Please refer to PERFORMANCE for some benchmarks and performance tuning.


	How to test yespower for proper operation.

On a Unix-like system, invoke "make check".  This will build and run a
program called "tests", and check its output against the supplied file
TESTS-OK.  If everything matches, the final line of output should be the
word "PASSED".

We do most of our testing on Linux systems with gcc.  The supplied
Makefile assumes that you use gcc.


	Alternate code versions and make targets.

Two implementations of yespower are included: reference and optimized.
By default, the optimized implementation is built.  Internally, the
optimized implementation uses conditional compilation to choose between
usage of various SIMD instruction sets where supported and scalar code.

The reference implementation is unoptimized and is very slow, but it has
simpler and shorter source code.  Its purpose is to provide a simple
human- and machine-readable specification that implementations intended
for actual use should be tested against.  It is deliberately mostly not
optimized, and it is not meant to be used in production.

Similarly to "make check", there's "make check-ref" to build and test
the reference implementation.  There's also "make ref" to build the
reference implementation and have the "benchmark" program use it.

"make clean" may need to be run between making different builds.


	How to integrate yespower in a program.

Although yespower.h provides several functions, chances are that you
will only need to use yespower_tls().  Please see the comment on this
function in yespower.h and its example usage in tests.c and benchmark.c,
including parameter sets requesting yescrypt 0.5 as used by certain
existing cryptocurrencies.

To integrate yespower in an altcoin based on Bitweb Core, you might
invoke yespower_tls() from either a maybe-new (depending on where you
fork from) CBlockHeader::GetPoWHash() (and invoke that where PoW is
needed like e.g. Litecoin does for scrypt) or CBlockHeader::GetHash()
(and implement caching for its return value like e.g. YACoin does for
scrypt).  Further detail on this (generating new genesis blocks, etc.)
is not yespower-specific and thus is not provided here.  Just like (and
even more so than) yespower itself, this guidance is provided as-is and
without guarantee of being correct and safe to follow.  You're supposed
to know what you're doing.


	Credits.

scrypt has been designed by Colin Percival.  yescrypt and yespower have
been designed by Solar Designer building upon scrypt.

The following other people and projects have also indirectly helped make
yespower what it is:

 - Bill Cox
 - Rich Felker
 - Anthony Ferrara
 - Christian Forler
 - Taylor Hornby
 - Dmitry Khovratovich
 - Samuel Neves
 - Marcos Simplicio
 - Ken T Takusagawa
 - Jakob Wenzel
 - Christian Winnerlein

 - DARPA Cyber Fast Track
 - Password Hashing Competition


	Contact info.

First, please check the yespower homepage for new versions, etc.:

    http://www.openwall.com/yespower/

If you have anything valuable to add or a non-trivial question to ask,
you may contact the maintainer of yespower at:

    Solar Designer <solar at openwall.com>


CHANGES
	Changes made between 1.0.0 (2018/07/12) and 1.0.1 (2019/06/30).

Fill the destination buffer with all set bits on error for fail-safety
of the caller's "< target" check in case the caller neglects to check
for errors.

Simplified SMix2 for its final invocation with Nloop=2 in yespower 0.5.

Revised the "XOR of yespower" tests to trigger duplicate index in the
last SMix2 invocation in yespower 0.5 for N=2048 with at least one of
the values of r being tested.  This is needed to test that a proper
kind of BlockMix is used in that special case, which would previously be
left untested.

Added x32 ABI support (x86-64 with 32-bit pointers).

Added a bit more detail to the README on caching of the computed PoW
hashes when integrating yespower in an altcoin based on Bitcoin Core.


tests.c
/*-
 * Copyright 2013-2018 Alexander Peslyak
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <stdio.h>
#include <string.h>

#include "yespower.h"

#undef TEST_PBKDF2_SHA256

#ifdef TEST_PBKDF2_SHA256
#include <assert.h>

#include "sha256.h"

static void print_PBKDF2_SHA256_raw(const char *passwd, size_t passwdlen,
    const char *salt, size_t saltlen, uint64_t c, size_t dkLen)
{
	uint8_t dk[64];
	size_t i;

	assert(dkLen <= sizeof(dk));

	/* XXX This prints the strings truncated at first NUL */
	printf("PBKDF2_SHA256(\"%s\", \"%s\", %llu, %llu) = ",
	    passwd, salt, (unsigned long long)c, (unsigned long long)dkLen);

	PBKDF2_SHA256((const uint8_t *) passwd, passwdlen,
	    (const uint8_t *) salt, saltlen, c, dk, dkLen);

	for (i = 0; i < dkLen; i++)
		printf("%02x%c", dk[i], i < dkLen - 1 ? ' ' : '\n');
}

static void print_PBKDF2_SHA256(const char *passwd,
    const char *salt, uint64_t c, size_t dkLen)
{
	print_PBKDF2_SHA256_raw(passwd, strlen(passwd), salt, strlen(salt), c,
	    dkLen);
}
#endif

static const char *pers_bsty_magic = "BSTY";

static void print_yespower(yespower_version_t version, uint32_t N, uint32_t r,
    const char *pers)
{
	yespower_params_t params = {
		.version = version,
		.N = N,
		.r = r,
		.pers = (const uint8_t *)pers,
		.perslen = pers ? strlen(pers) : 0
	};
	uint8_t src[80];
	yespower_binary_t dst;
	size_t i;

	const char *q = (pers && pers != pers_bsty_magic) ? "\"": "";
	printf("yespower(%u, %u, %u, %s%s%s) = ", (unsigned int)version, N, r,
	    q, pers ? pers : "NULL", q);

	for (i = 0; i < sizeof(src); i++)
		src[i] = i * 3;

	if (pers == pers_bsty_magic) {
		params.pers = src;
		params.perslen = sizeof(src);
	}

	if (yespower_tls(src, sizeof(src), &params, &dst)) {
		puts("FAILED");
		return;
	}

	for (i = 0; i < sizeof(dst); i++)
		printf("%02x%c", dst.uc[i], i < sizeof(dst) - 1 ? ' ' : '\n');
}

static void print_yespower_loop(yespower_version_t version, const char *pers)
{
	uint32_t N, r;
	uint8_t src[80];
	yespower_binary_t dst, xor = {{0}};
	size_t i;

	printf("XOR of yespower(%u, ...) = ", (unsigned int)version);

	/*
	 * This value of src is chosen to trigger duplicate index in the last
	 * SMix2 invocation in yespower 0.5 for N=2048 with at least one of the
	 * values of r below.  This is needed to test that a non-save version
	 * of BlockMix is used in that special case.  Most other values of src
	 * would leave this untested.
	 */
	src[0] = 43;
	for (i = 1; i < sizeof(src); i++)
		src[i] = i * 3;

	for (N = 1024; N <= 4096; N <<= 1) {
		for (r = 8; r <= 32; r++) {
			yespower_params_t params = {
				.version = version,
				.N = N,
				.r = r,
				.pers = (const uint8_t *)pers,
				.perslen = pers ? strlen(pers) : 0
			};
			if (yespower_tls(src, sizeof(src), &params, &dst)) {
				puts("FAILED");
				return;
			}
			for (i = 0; i < sizeof(xor); i++)
				xor.uc[i] ^= dst.uc[i];
		}
	}

	for (i = 0; i < sizeof(xor); i++)
		printf("%02x%c", xor.uc[i], i < sizeof(xor) - 1 ? ' ' : '\n');
}

int main(void)
{
	setvbuf(stdout, NULL, _IOLBF, 0);

#ifdef TEST_PBKDF2_SHA256
	print_PBKDF2_SHA256("password", "salt", 1, 20);
	print_PBKDF2_SHA256("password", "salt", 2, 20);
	print_PBKDF2_SHA256("password", "salt", 4096, 20);
	print_PBKDF2_SHA256("password", "salt", 16777216, 20);
	print_PBKDF2_SHA256("passwordPASSWORDpassword",
	    "saltSALTsaltSALTsaltSALTsaltSALTsalt", 4096, 25);
	print_PBKDF2_SHA256_raw("pass\0word", 9, "sa\0lt", 5, 4096, 16);
#if 0
	print_PBKDF2_SHA256("password", "salt", 1, 32);
	print_PBKDF2_SHA256("password", "salt", 2, 32);
	print_PBKDF2_SHA256("password", "salt", 4096, 32);
	print_PBKDF2_SHA256("password", "salt", 16777216, 32);
	print_PBKDF2_SHA256("passwordPASSWORDpassword",
	    "saltSALTsaltSALTsaltSALTsaltSALTsalt", 4096, 40);
	print_PBKDF2_SHA256("password", "salt", 4096, 16);
	print_PBKDF2_SHA256("password", "salt", 1, 20);
	print_PBKDF2_SHA256("password", "salt", 2, 20);
	print_PBKDF2_SHA256("password", "salt", 4096, 20);
	print_PBKDF2_SHA256("password", "salt", 16777216, 20);
	print_PBKDF2_SHA256("password", "salt", 4096, 25);
	print_PBKDF2_SHA256("password", "salt", 4096, 16);
#endif
#endif

	print_yespower(YESPOWER_0_5, 2048, 8, "Client Key"); /* yescrypt 0.5 */
	print_yespower(YESPOWER_0_5, 2048, 8, pers_bsty_magic); /* BSTY */
	print_yespower(YESPOWER_0_5, 4096, 16, "Client Key"); /* Cryply */
	print_yespower(YESPOWER_0_5, 4096, 24, "Jagaricoin");
	print_yespower(YESPOWER_0_5, 4096, 32, "WaviBanana");
	print_yespower(YESPOWER_0_5, 2048, 32, "Client Key");
	print_yespower(YESPOWER_0_5, 1024, 32, "Client Key");

	print_yespower(YESPOWER_0_5, 2048, 8, NULL); /* no personality */

	print_yespower(YESPOWER_1_0, 2048, 8, NULL);
	print_yespower(YESPOWER_1_0, 4096, 16, NULL);
	print_yespower(YESPOWER_1_0, 4096, 32, NULL);
	print_yespower(YESPOWER_1_0, 2048, 32, NULL);
	print_yespower(YESPOWER_1_0, 1024, 32, NULL);

	print_yespower(YESPOWER_1_0, 1024, 32, "personality test");

	print_yespower_loop(YESPOWER_0_5, "Client Key");
	print_yespower_loop(YESPOWER_1_0, NULL);

	return 0;
}


benchmark.c
/*-
 * Copyright 2013-2018 Alexander Peslyak
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <stdio.h>
#include <stdlib.h> /* for atoi() */
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <sys/times.h>
#include <sched.h>

#include "yespower.h"

#ifdef _OPENMP
#include <omp.h>

#define NSAVE 1000

static uint64_t time_us(void)
{
	struct timespec t;
#ifdef CLOCK_MONOTONIC_RAW
	if (clock_gettime(CLOCK_MONOTONIC_RAW, &t))
		return 0;
#else
	if (clock_gettime(CLOCK_MONOTONIC, &t))
		return 0;
#endif
	return 1 + (uint64_t)t.tv_sec * 1000000 + t.tv_nsec / 1000;
}
#endif

int main(int argc, const char * const *argv)
{
	yespower_params_t params = {
		.version = YESPOWER_0_5,
		.N = 2048,
		.r = 8,
		.pers = (const uint8_t *)"Client Key",
		.perslen = 10
	};

	if (argc > 1)
		params.version = atoi(argv[1]);
	if (argc > 2)
		params.N = atoi(argv[2]);
	if (argc > 3)
		params.r = atoi(argv[3]);

	printf("version=%.1f N=%u r=%u\n",
	    params.version * 0.1, params.N, params.r);

	printf("Will use %.2f KiB RAM\n", 0.125 * params.N * params.r);

	static __thread union {
		uint8_t u8[80];
		uint32_t u32[20];
	} src;
	yespower_binary_t dst;
	unsigned int i;

	for (i = 0; i < sizeof(src); i++)
		src.u8[i] = i * 3;

	if (yespower_tls(src.u8, sizeof(src), &params, &dst)) {
		puts("FAILED");
		return 1;
	}

	for (i = 0; i < sizeof(dst); i++)
		printf("%02x%c", dst.uc[i], i < sizeof(dst) - 1 ? ' ' : '\n');

	puts("Benchmarking 1 thread ...");

	clock_t clk_tck = sysconf(_SC_CLK_TCK);
	struct tms start_tms, end_tms;
	clock_t start = times(&start_tms), end;
	unsigned int n;
	unsigned long long count;
#ifdef _OPENMP
	yespower_binary_t save[NSAVE];
	unsigned int nsave = 0;
#endif
	uint32_t seed = start * 1812433253U;

	n = 1;
	count = 0;
	do {
		for (i = 0; i < n; i++) {
			yespower_binary_t *p = &dst;
#ifdef _OPENMP
			if (nsave < NSAVE)
				p = &save[nsave++];
#endif
			src.u32[19] = seed + (count + i);
			if (yespower_tls(src.u8, sizeof(src), &params, p)) {
				puts("FAILED");
				return 1;
			}
		}
		count += n;

		end = times(&end_tms);
		n <<= 1;
	} while (end - start < clk_tck * 2);

	clock_t start_v = start_tms.tms_utime + start_tms.tms_stime +
	    start_tms.tms_cutime + start_tms.tms_cstime;
	clock_t end_v = end_tms.tms_utime + end_tms.tms_stime +
	    end_tms.tms_cutime + end_tms.tms_cstime;

	printf("%llu H/s real, %llu H/s virtual "
	    "(%llu hashes in %.2f seconds)\n",
	    count * clk_tck / (end - start),
	    count * clk_tck / (end_v - start_v),
	    count, (double)(end - start) / clk_tck);

	for (i = 0; i < nsave; i++) {
		unsigned int j;
		for (j = i + 1; j < nsave; j++) {
			unsigned int k = 8;
			if (!memcmp(&save[i], &save[j], k)) {
				printf("%u-byte collision(s) detected\n", k);
				i = nsave; break;
			}
		}
	}

#ifdef _OPENMP
	unsigned int nt = omp_get_max_threads();

	printf("Benchmarking %u thread%s ...\n",
	    nt, nt == 1 ? "" : "s");

	typedef struct {
		uint64_t min, max, total;
	} thread_data_s;
	union {
		thread_data_s s;
		uint8_t cachelines[2][64]; /* avoid false sharing */
	} thread_data[nt]; /* tricky to align this when on stack */

	unsigned int t;
	for (t = 0; t < nt; t++) {
		thread_data_s *td = &thread_data[t].s;
		td->min = ~(uint64_t)0; td->max = 0; td->total = 0;
	}

	unsigned long long count1 = count, count_restart = 0;

	if (!geteuid()) {
		puts("Running as root, so trying to set SCHED_RR");
#pragma omp parallel
		{
			struct sched_param param = { .sched_priority = 1 };
			if (sched_setscheduler(getpid(), SCHED_RR, &param))
				perror("sched_setscheduler");
		}
	}

	start = times(&start_tms);

	n = count * omp_get_max_threads();
	count = 0;
	do {
#pragma omp parallel for default(none) copyin(src) private(i, dst) shared(n, thread_data, params, seed, count, save, nsave)
		for (i = 0; i < n; i++) {
			unsigned int j = count + i;

			src.u32[19] = seed + j;

			uint64_t start1 = time_us();

			if (yespower_tls(src.u8, sizeof(src), &params, &dst)) {
#pragma omp critical
				puts("FAILED");
			}

			uint64_t end1 = time_us();
			if (end1 < start1)
				end1 = start1;
			uint64_t diff1 = end1 - start1;

			thread_data_s *td = &thread_data[omp_get_thread_num()].s;
			td->total += diff1;
			if (diff1 < td->min)
				td->min = diff1;
			if (diff1 > td->max)
				td->max = diff1;

#ifdef _OPENMP
			if (j < nsave && memcmp(&save[j], &dst, sizeof(dst))) {
#pragma omp critical
				printf("Mismatch at %u\n", j);
			}
#endif
		}

		count += n;
		if ((count - n) < count1 && count >= count1) {
/* Disregard our repeat of single thread's results (could be partially cached
 * by same core, but OTOH other cores not yet warmed up to full clock rate). */
			start = times(&start_tms);
			count_restart = count;
			for (t = 0; t < nt; t++) {
				thread_data_s *td = &thread_data[t].s;
				td->min = ~(uint64_t)0; td->max = 0; td->total = 0;
			}
		} else {
			n <<= 1;
		}

		end = times(&end_tms);
	} while (end - start < clk_tck);

	if (!count_restart)
		puts("Didn't reach single-thread's hash count");
	count -= count_restart;

	start_v = start_tms.tms_utime + start_tms.tms_stime +
	    start_tms.tms_cutime + start_tms.tms_cstime;
	end_v = end_tms.tms_utime + end_tms.tms_stime +
	    end_tms.tms_cutime + end_tms.tms_cstime;

	printf("%llu H/s real, %llu H/s virtual "
	    "(%llu hashes in %.2f seconds)\n",
	    count * clk_tck / (end - start),
	    count * clk_tck / (end_v - start_v),
	    count, (double)(end - start) / clk_tck);

	uint64_t min = ~(uint64_t)0, max = 0, total = 0;
	for (t = 0; t < nt; t++) {
		thread_data_s *td = &thread_data[t].s;
		total += td->total;
		if (td->min < min)
			min = td->min;
		if (td->max > max)
			max = td->max;
	}
	printf("min %.3f ms, avg %.3f ms, max %.3f ms\n",
		min / 1000.0, total / 1000.0 / count, max / 1000.0);
#endif

	return 0;
}
